### ISSUE TITLE
When called via a script, hexdump pipeline always times out

### ISSUE THREAD

**Issue by cheeseandcereal** [CapitalF/gridfan#2]
I set up the script and was able to get it working most of the time when manually using the script over the cli, but whenever I try to call gridfan from within another shell script, any command sent to the controller times out without fail, even if I include a 'gridfan init' right above any other commands.

This happens regardless of the command I'm trying to send (a ping, show, or set), and is only remedied when I go back to the cli and manually run a 'gridfan init'.

The debug output whenever I run a script just shows the hexdump pipeline timing out every time. Not sure what the issue is, but I can help debug if you have any ideas.

**Comment by CapitalF** [comment]
Do a "gridfan -D ping" from your regular bash shell and then from within your script and make sure there isn't a difference in the stty settings.

Please post the result of "gridfan -D ping", which I expect will show a failure, from within your script here so that I can look at it.

I may want to look at your script's env and what it's exporting. Could it be exporting something that interferes with gridfan?

Tell me about this script which is calling gridfan. Is it running as a user with permissions to access the $GRID_DEV ?

**Comment by cheeseandcereal** [comment]
gridfan -D ping from cli:
```
DEBUG: f_sendhex: input "IN_BYTES=C0", "IN_NUM_BYTES=1".
DEBUG: f_setup_dev: stty setting parameters to the device: "4800 raw -echo -echoe -echok".
DEBUG: f_xmit: Flushing device output before transmission.
DEBUG: f_xmit: Sending XMIT_CMD "\xC0" to GRID_DEV "/dev/GridPlus0".
DEBUG: f_receive: Waiting for hexdump pipeline to finish. "PID=16232", "WAIT_COUNT=1".
DEBUG: f_sendhex: f_receive exit code "0".
DEBUG: f_sendhex: Received OUT_NUM_BYTES="1", OUT_BYTES="21".
DEBUG: f_ping: f_sendhex returned "0".

Success! Controller responded to our ping.
```
This is the simplest script I could write which consistently reproduces the error. I'm calling this script as my user which previously successfully did the gridfan -D ping from above. If I remove the line where I attempt to set a fan speed, the ping works fine, but after that line, the ping will always fail (it doesn't successfully set the fan speed either, the same timeout occurs for that line as well)

```
#!/bin/bash
/home/adam/gridv2/gridfan init
/home/adam/gridv2/gridfan set fans 1 speed 80
/home/adam/gridv2/gridfan -D ping
```
-D ping Output (from above script):
```
DEBUG: f_sendhex: input "IN_BYTES=C0", "IN_NUM_BYTES=1".
DEBUG: f_setup_dev: stty setting parameters to the device: "4800 raw -echo -echoe -echok".
DEBUG: f_xmit: Flushing device output before transmission.
DEBUG: f_xmit: Sending XMIT_CMD "\xC0" to GRID_DEV "/dev/GridPlus0".
DEBUG: f_receive: Waiting for hexdump pipeline to finish. "PID=20431", "WAIT_COUNT=1".
DEBUG: f_receive: Waiting for hexdump pipeline to finish. "PID=20431", "WAIT_COUNT=2".
    <EXCLUDED UNNECESSARY WAITING LINES>
DEBUG: f_receive: Waiting for hexdump pipeline to finish. "PID=20431", "WAIT_COUNT=99".
DEBUG: f_receive: Waiting for hexdump pipeline to finish. "PID=20431", "WAIT_COUNT=100".
DEBUG: f_receive: Killing hexdump pipeline due to timeout.
DEBUG: f_receive: hexdump pipeline killed.
DEBUG: f_sendhex: f_receive exit code "9".
DEBUG: f_sendhex: Received OUT_NUM_BYTES="0", OUT_BYTES="".
DEBUG: f_ping: f_sendhex returned "9".

ERROR: Command to controller timed out. No reply received. (Do you need to "gridfan init"?)
```

**Comment by CapitalF** [comment]
I can reproduce your problem, so it's not just you.

It might be a day or two before I can look into this further, but I'll get back to it.

**Comment by CapitalF** [comment]
Try setting you fan speeds to anything but 80 or 85.

Looks like there might be a problem with the hex codes for 80 and 85%.

**Comment by cheeseandcereal** [comment]
Yup, seems like that works. I guess I never tried setting 80 or 85% manually on the command line. 

Thanks.

**Comment by CapitalF** [comment]
I will do a full test of all speeds later after I figure out what the problem with the byte codes are, to make sure all of them work.

Thanks for finding and reporting this!

I will update once I get this figured out. Might be a week though.

**Comment by CapitalF** [comment]
This should be fixed in version 1.5, which I just uploaded.

I had been playing fast-and-lose with the Grid device and that's easy since I wrote this in bash, which is non-ideal because of issues exactly like this.

For some reason, writing command "44 06 C0 00 00 0a 00" to the Grid fails if I've not already opened up the device for reading. This probably is related to the fact that 0a is the ASCII LF character.

Previously I would blindly send the command to the device and then later read the reply. Now we properly open the device for reading and writing at the same time, send the command, and get the reply back.
